# Using Bank Marketing dataset we need to predict whether the customer will subscribe to our servises. Subscription is our target

- 1. What is the accuracy when you use a Random Forest?
- 2. What is the accuracy when you use an SVM? How does changing the model affect the accuracy?
- 3. Find the top 4 important factors by using Random Forest. What is you inference?
- 4. Which evaluation metric is important and why?




In [2]:
#Importing the libraries
import pandas as pd

In [3]:
bank_marketing = pd.read_excel('F:\Skillenable\Data frames\Bank marketing.xlsx')
bank_marketing.head()

Age            Job Marital Status  Account Balance  Own House  \
0   59     unemployed        married                0          0   
1   36         Others        married             3057          0   
2   47    blue-collar       divorced              126          1   
3   43     management       divorced              388          1   
4   34  self-employed         single              462          0   

   Personal Loan  No of campaigns  Subscription  
0              0                0             0  
1              0                0             1  
2              0                0             1  
3              0                0             1  
4              0                0             1

In [4]:
bank_marketing.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 8 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Age              4521 non-null   int64 
 1   Job              4521 non-null   object
 2   Marital Status   4521 non-null   object
 3   Account Balance  4521 non-null   int64 
 4   Own House        4521 non-null   int64 
 5   Personal Loan    4521 non-null   int64 
 6   No of campaigns  4521 non-null   int64 
 7   Subscription     4521 non-null   int64 
dtypes: int64(6), object(2)
memory usage: 282.7+ KB


In [5]:
bank_marketing.isnull().sum()   

Age                0
Job                0
Marital Status     0
Account Balance    0
Own House          0
Personal Loan      0
No of campaigns    0
Subscription       0
dtype: int64

In [6]:

bank_marketing.shape

(4521, 8)

In [7]:
bank_marketing.shape[0]  

4521

In [8]:
# crete dummy data for catagorical values
cat_col = ['Job', 'Marital Status']
cat_col

['Job', 'Marital Status']

In [10]:
bank_dummy = pd.get_dummies(bank_marketing, columns = cat_col, drop_first= True)
bank_dummy

Age  Account Balance  Own House  Personal Loan  No of campaigns  \
0      59                0          0              0                0   
1      36             3057          0              0                0   
2      47              126          1              0                0   
3      43              388          1              0                0   
4      34              462          0              0                0   
...   ...              ...        ...            ...              ...   
4516   58                0          1              0                0   
4517   38                0          1              0                0   
4518   53             2627          1              1                0   
4519   49              371          1              0                0   
4520   35              280          1              1                0   

      Subscription  Job_blue-collar  Job_management  Job_self-employed  \
0                0                0               0                  0   
1                1                0               0                  0   
2                1                1               0                  0   
3                1                0               1                  0   
4                1                0               0                  1   
...            ...              ...             ...                ...   
4516             0                1               0                  0   
4517             0                0               0                  0   
4518             0                0               0                  1   
4519             0                1               0                  0   
4520             0                1               0                  0   

      Job_unemployed  Marital Status_married  Marital Status_single  
0                  1                       1                      0  
1                  0                       1                      0  
2                  0                       0                      0  
3                  0                       0                      0  
4                  0                       0                      1  
...              ...                     ...                    ...  
4516               0                       1                      0  
4517               0                       1                      0  
4518               0                       1                      0  
4519               0                       1                      0  
4520               0                       1                      0  

[4521 rows x 12 columns]

In [11]:
# create input , output 
Y = bank_dummy[['Subscription']]
X = bank_dummy.drop(columns=['Subscription'])

In [12]:
# spliting data into test and train

from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,random_state=47)
len(X_train),len(X_test),len(Y_train),len(Y_test)

(3616, 905, 3616, 905)

## 1) What is the accuracy when you use as Random forest?

In [19]:
# creating random forest model
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators=600,random_state=1234)
rf = model.fit(X_train,Y_train)


C:\Users\arun\AppData\Local\Temp/ipykernel_9220/2587773235.py:5: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf = model.fit(X_train,Y_train)


In [20]:
# Predicting the model 
Y_test

Subscription  Prediction
1034             0           0
3673             0           0
1004             0           0
3489             0           0
682              0           0
...            ...         ...
1326             0           0
2012             0           0
2519             0           0
755              0           0
2917             0           0

[905 rows x 2 columns]

In [21]:
Y_test['Prediction'] = model.predict(X_test)
Y_test

Subscription  Prediction
1034             0           0
3673             0           0
1004             0           0
3489             0           0
682              0           0
...            ...         ...
1326             0           0
2012             0           0
2519             0           0
755              0           0
2917             0           0

[905 rows x 2 columns]

In [22]:
from sklearn.metrics import confusion_matrix, accuracy_score
print(confusion_matrix(Y_test['Subscription'],Y_test['Prediction']))

[[762  27]
 [106  10]]


In [23]:
print(accuracy_score(Y_test['Subscription'],Y_test['Prediction']))


0.8530386740331491


### the accuracy of Forest tree is 85.31%

## 3) Find the top 4 important factors by using Random Forest. What is your inference?

In [24]:
model.feature_importances_

array([0.29758396, 0.47396068, 0.02760333, 0.01674481, 0.0927109 ,
       0.01759889, 0.02183623, 0.00968473, 0.00761587, 0.01961593,
       0.01504467])

In [25]:
forest_importance = pd.Series(model.feature_importances_,index=X.columns)
forest_importance.sort_values(ascending=False).head(10)

Account Balance           0.473961
Age                       0.297584
No of campaigns           0.092711
Own House                 0.027603
Job_management            0.021836
Marital Status_married    0.019616
Job_blue-collar           0.017599
Personal Loan             0.016745
Marital Status_single     0.015045
Job_self-employed         0.009685
dtype: float64

## 2) What is the accuracy when you use an SVM? How does changing the model affect the accuracy?


In [26]:
# building SVM 
from sklearn.svm import SVC
model1 = SVC(random_state=47)

In [27]:
svc_model = model1.fit(X_train,Y_train)  


C:\Users\arun\anaconda3\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [28]:
# prediction
Y_test['Prediction'] = svc_model.predict(X_test)
Y_test


Subscription  Prediction
1034             0           0
3673             0           0
1004             0           0
3489             0           0
682              0           0
...            ...         ...
1326             0           0
2012             0           0
2519             0           0
755              0           0
2917             0           0

[905 rows x 2 columns]

In [29]:
# confusion metrix
from sklearn.metrics import confusion_matrix,accuracy_score
print(confusion_matrix(Y_test['Subscription'],Y_test['Prediction']))

[[789   0]
 [116   0]]


In [30]:
# accuracy
print(accuracy_score(Y_test['Subscription'],Y_test['Prediction']))


0.8718232044198895


###  Accuracy of SVM is 87.18 %; accuracy of SVM is higher than Random Forest 

## 4) Which evaluation metrics is important and why?

- As per the result SVM gave higher accuracy than random forest model. 
- For our case SVM is better. 

